### DISCUSSION TITLE
Is there any way to create your own tool texture / special face type (like CLIP and SKIP)

### DISCUSSION THREAD

**Discussion by aslanhudajev** [func-godot/func_godot_plugin#70] [Q&A]
Hi, the title of this thread pretty much says it all.
Is there any way to create your own tool textures, like the clip and skip textures?

The Qodot manual mentions tool textures, but does not specify how one would have to go about creating their own in Godot: https://qodotplugin.github.io/docs/tips-for-veterans.html#tool-textures.
And as far as I know the func_godot manual doesn't mention them at all (outside of the fact that clip and skip exist).

Here's an example of how tool textures could be utilized:
https://developer.valvesoftware.com/wiki/Tool_textures_(Source)

**Comment by sinewavey** [comment]
Yes.

Take a look at the [manual section just before _Automatic PBR Generation_](https://func-godot.github.io/func_godot_docs/FuncGodot%20Manual/pages/guide_textures.html):

> FuncGodot's first step is to search the Base Texture Dir for a prebuilt material as defined by the brush face's **texture name** in the **format defined by Material File Extension.**
>
> If it does not find this prebuilt material, it will search the Base Texture Dir for the Texture2D resource by the brush face's texture name using any one of the extensions defined by Texture File Extensions. If it finds the Texture2D, it will create a copy of the Default Material and apply the Texture2D as the duplicated material's albedo.
>
> If it does not find the Texture2D, then it will use the default_texture.png found in the addons/func_godot/textures/ folder. 

What I've highlighted, rephrased:

"If you name a texture _my_tool.png_, and name a material _my_tool.tres_, the material will be applied first before creating a new texture2D with this image for albedo"

**Reply by sinewavey** [reply]
Do note: it's probably better to create desired behavior with entities, but provide them with textures that can map to another material.

Otherwise, you'll need to make changes yourself to the codebase to detect these specific textures and direct them to specific actions, like how the Origin brush texture is handled.
For example: what's a weapon clip vs a player clip? It's unanswerable without some specific game context. What that means in Godot is dependent on *your game*.

**Comment by RhapsodyInGeek** [comment]
> The Qodot manual mentions tool textures, but does not specify how one would have to go about creating their own in Godot:
> https://qodotplugin.github.io/docs/tips-for-veterans.html#tool-textures.
> And as far as I know the func_godot manual doesn't mention them at all (outside of the fact that clip and skip exist).

The section you linked also tells you the only tool textures it uses are Skip and Clip. You don't really create your own tool textures in a way that FuncGodot can utilize at build time, nor does it suggest you can create your own that affect the build process. The Trigger texture suggestion is gives is "for the sake of organization and clarity" and has nothing to do with actual functionality.

FuncGodot has support for Skip, Clip, and Origin, though Skip and Clip function a bit differently than they do in id Tech based games (including GoldSrc and Source). [All of these are described in the FuncGodot Manual](https://func-godot.github.io/func_godot_docs/FuncGodot%20Manual/pages/guide_textures.html#ClipAndSkip).

> Here's an example of how tool textures could be utilized:
> https://developer.valvesoftware.com/wiki/Tool_textures_(Source)

Most of that is either already covered by Skip and Clip, or out of scope for FuncGodot's design. There's no way for FuncGodot to know what is an NPC and what is a Player in your game, or whether you even divide NPCs and Players up that way. FuncGodot doesn't care about what a projectile is.

Quake has a distinction between Player Collision and General Entity Collision. Half-Life further provides some further distinctions... about 4 or 5 I think? Godot provides 64 different collision layers that can be mixed and matched. If you want to provide that distinction, FuncGodot allows you to create entities that can cover specific layers by default in the SolidClass definition, or you can add a class property that sets the collision layers and masks for your solid geometry.

For your trigger entities, FuncGodot allows you to specify whether to build a visual mesh or not, or whether to build collision.

If you need further tool texture support, you can technically build your own via the new Mesh Metadata features we added, or through sinewavey's suggestions of [creating textures that map to other materials during build](https://func-godot.github.io/func_godot_docs/FuncGodot%20Manual/pages/guide_textures.html).

The first question you should ask yourself though is "why do I need extra tool textures?" Figure out the problems you're actually trying to solve before trying to figure out the features you need.
